capreolus/sampler/__init__.py

```diff
@@ -0,0 +1,97 @@
+import random
+import torch.utils.data
+
+from capreolus.utils.loginit import get_logger
+
+
+logger = get_logger(__name__)
+
+
+class TrainDataset(torch.utils.data.IterableDataset):
+    """
+    Samples training data. Intended to be used with a pytorch DataLoader
+    """
+
+    def __init__(self, search_run, benchmark, extractor):
+        self.search_run = search_run
+        self.benchmark = benchmark
+        self.extractor = extractor
+
+    def __iter__(self):
+        """
+        Returns: Triplets of the form (query_feature, posdoc_feature, negdoc_feature)
+        """
+        extractor = self.extractor
+        benchmark = self.benchmark
+        search_run = self.search_run
+
+        train_qids = benchmark.folds[benchmark.cfg["fold"]]["train_qids"]
+        is_rundocsonly = benchmark.cfg["rundocsonly"]
+        if is_rundocsonly:
+            # Create qrels from only those docs that showed up in the search results
+            qrels = {
+                qid: {docid: label for docid, label in benchmark.qrels[qid].items() if docid in search_run[qid]}
+                for qid in train_qids
+            }
+        else:
+            qrels = benchmark.qrels
+
+        qid_to_reldocs = {qid: [docid for docid, label in qrels[qid].items() if label > 0] for qid in qrels}
+        qid_to_negdocs = {qid: [docid for docid, label in qrels[qid].items() if label <= 0] for qid in qrels}
+
+        # Remove the qids that does not have relevant/irrelevant docs associated with it
+        valid_qids = [qid for qid in train_qids if qid_to_reldocs.get(qid) and qid_to_negdocs.get(qid)]
+
+        # Convert each query and doc id to the corresponding feature/embedding and yield
+        def genf():
+            while True:
+                # random.shuffle(valid_qids)
```

Comment:
I should set the seed before calling a shuffle right? 
We can have a `self.iteration` variable attached to the dataset class, and then increment it on every yield. Then the seed would be set here as `random.seed(self.iteration)`. The problem here is that `self.iteration` would be the number of times yield has been called - i.e `num_iterations` * `batch_size` - since we increment `self.iterations` on every yield. Not sure if this would be a problem.